### PR TITLE
Read model type from R (attempt number 2) [PLT-1174]

### PR DESCRIPTION
**Pull request by kwxm**
This is a continuation of #5841.  It
  * Improves R/Haskell interface in the code that generates the cost model, simplifying things considerably.
  * Moves the memory cost models into a separate file.
  * Removes some hard-coded numbers from the Haskell code and generates them in the R code instead.
  * Generalises the costing function infrastructure a bit: for example a number of models involved special types that were really the coeffcients of linear functions but I've now replaced those all with a single type of linear function.  This should make it easier to change the model type if we need to, for example by replacing a linear function with a logarithmic one.  This has led to some more nesting in the costing function types which could possibly slow things down a bit.  I'll run the benchmarks to check.
  * The changes above have also changed the JSON nesting, but they haven't changed the order of the parameters, so cost model updates shouldn't be affected.

These changes will lead to some small differences in the numbers in the JSON file, but I haven't updated it yet.  It's probably better to wait for a full cost model update to do that.  The one exception is the cost model for `verifyEd25519Signature`, which was based on the wrong variable.  I've taken the opportunity to correct that. The function is very fast anyway, so the differences weren't very significant: the correction has actually made the benchmarks _cheaper_. [This change has now been reverted until some other pending PRs which improve configurability have been merged.].

I have some more improvements in mind for the core costing function code, but those will probably change the order of the cost model parameters so I'll leave it for a later PR.

### CODE REVIEW

**Comment by kwxm** [comment]
I was mystified by the apparent speedup in `nofib`, but #5869 suggests that the faster results are also what happens in master.  Presumably some previous PR sped things up and I merged that into this branch at some point.

**Review by effectfully — COMMENTED** [review]
A lot of that is above my head, but here are some comments.

Generally, I think we're moving in the opposite direction of what I consider to be the right approach: making static info (shapes of costing functions) into static code inside of builtins where it's more convenient to review, maintain etc and where it would be faster too.

Is it in theory possible to supply R code with the shapes of costing functions and parse it there as opposed to doing it the opposite way?

File: plutus-metatheory/src/Cost/Model.lagda.md
Comment (kwxm) [inline plutus-metatheory/src/Cost/Model.lagda.md:1]: I was fixing something else in the Agda code and trimmed lots of trailing whitespace while I was at it.

**Comment by kwxm** [comment]
> Is it in theory possible to supply R code with the shapes of costing functions and parse it there as opposed to doing it the opposite way?

I think not. The R code does quite a lot of other stuff, like [selecting subsets of the data](https://github.com/IntersectMBO/plutus/blob/757533620b3e66ec1777728175661b9ea674e7ef/plutus-core/cost-model/data/models.R#L420) to fit a model to.  We'd have to tell it all that as well and then we're moving towards a full-scale DSL for interacting with R.  I think we'd end up translating much of the R code into more complicated Haskell.

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostingFun/Core.hs [hunk 1]
```diff
@@ -260,13 +256,10 @@ evaluateOneVariableQuadraticFunction
...
-data ModelAddedSizes = ModelAddedSizes
-    { modelAddedSizesIntercept :: Intercept
-    , modelAddedSizesSlope     :: Slope
-    } deriving stock (Show, Eq, Generic, Lift)
-    deriving anyclass (NFData)
-
+-- FIXME: we could use ModelConstantOrOneArgument for
+-- ModelTwoArgumentsSubtractedSizes instead, but that would change the order of
+-- the cost model parameters since the minimum value would come first instead of
+-- last.
```
[the remark this answers is not in the document]
Comment (kwxm) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostingFun/Core.hs:262]: > Does anything relies on them existing?

I think not, and I'd like to generalise all of this stuff, but as it says that would change the order of the cost model parameters and we don't want to do that at the moment.  We have to worry about backwards compatibility as well.  In this PR I'm trying to improve things as much as I can without breaking anything else. Once we have more configurability wrt PVs and so on we can start to make some other changes.

File: plutus-core/cost-model/create-cost-model/CreateBuiltinCostModel.hs [hunk 2]
```diff
@@ -130,860 +117,263 @@ builtinCostModelNames = BuiltinCostModelBase
...
-   however we can still use that to model the constant cost inferred in the R
-   code.
+  cpuModels :: BuiltinCostModelBase (Const (SomeSEXP s)) <- costModelsR bmfile rfile
+  let getParams
+        :: (SomeSEXP s -> R s model)
+        -> (forall f. BuiltinCostModelBase f -> f model)
+        -> R s (CostingFun model)
+      getParams readCF param = do
+        let memModel = getId $ param builtinMemoryModels
+        cpuModel <- readCF $ getConst $ param cpuModels
```
[the remark this answers is not in the document]
Comment (kwxm) [reply plutus-core/cost-model/create-cost-model/CreateBuiltinCostModel.hs:149]: That's the way it's always been, and that's been OK because memory models don't change too much. Also we're not  doing any analysis to infer the memory models, just making them up based on what we know about the implemenations of the builtins.  What we could do is have a separate (versioned) JSON file for the memory models and read them from there: you'd have to write the JSON by hand though, which might be a bit fragile.  Maybe we could use some other format (which is essentially what we have in the Haskell now anyway). Actually maybe it's better just to keep them in `buikltinCostModel.json` anyway since if we eventually have mulitple versions of that it'd be easier to manage things if the memory models are in the same place as the CPU models.

**Comment by effectfully** [comment]
>  We'd have to tell it all that as well and then we're moving towards a full-scale DSL for interacting with R. I think we'd end up translating much of the R code into more complicated Haskell.

Perhaps I'm misunderstanding it, but I think all I need is for R to take a string or whatever and parse it, so that this:

```
    bls12_381_G2_negModel            <- constantModel ("Bls12_381_G2_neg")
    bls12_381_G2_scalarMulModel      <- linearInX     ("Bls12_381_G2_scalarMul")
    bls12_381_G2_equalModel          <- constantModel ("Bls12_381_G2_equal")
    bls12_381_G2_hashToGroupModel    <- linearInX     ("Bls12_381_G2_hashToGroup")
    bls12_381_G2_compressModel       <- constantModel ("Bls12_381_G2_compress")
```

doesn't have `constantModel`, `linearInX` etc hardcoded and instead I can hardcode them in the Haskell code and provide the R code with a string representing the name of the model. All I want is to be able to choose the appropriate shape of the costing function for each builtin right in the code for builtins where all the versioning logic is.

I'm not at all confident that would work on the Haskell side, so I'm asking if it can work on the R side, 'cause if not, then I should just give up on that dream (or consider returning the kind of double-hardcoding that you just removed, which you probably wouldn't approve of haha).

So basically the question is: do you think it's not possible for the Haskell code to know that `bls12_381_G2_equal` has a `constantModel`? I don't care about any coefficients, it's fine to supply those dynamically, or how the data is filtered etc, I just wanna to be able to tell GHC "depending on the protocol version the costing function here is either constant or linear".

**Comment by kwxm** [comment]
> So basically the question is: do you think it's not possible for the Haskell code to know that `bls12_381_G2_equal` has a `constantModel`? I don't care about any coefficients, it's fine to supply those dynamically, or how the data is filtered etc, I just wanna to be able to tell GHC "depending on the protocol version the costing function here is either constant or linear".

That seems a bit tricky.  We could probably do it for constant models, but the problem is that many of the models are quite precisely tailored to a particular builtin: you'd end up with things like
```
  divideIntegerModel <- constantAboveDiagonalMultipliedSizesBelowOrOnDiagonalModel ("DivideInteger")
```
Maybe we could make the type nested or something, but then we're kind of back to the DSL.

I'm also keen to keep the use of `inline-r` to a minimum: I'm not sure how well it's maintained and there may be a danger that it'll stop working altogether at some point.  I think we already got stuck with some patched version for a long time.  It seems better to let R do all of the stuff that it's good at and return a single object to Haskell that you query to extract the stuff you need.  Coming up with a cost model in the first place can be quite an interactive process where you work with the data inside R and fit different kinds of models to it and see how well they compare with reality, and having all the code be self-contained is helpful with that.  I even considered going completely in the other direction from what you're asking and just getting the R code to write the JSON file directly, which I think wouldn't be too hard.  Anyway, I'm not particularly inclined to embark on a major reworking of all this stuff at the moment: this PR's an attempt to make things a little more flexible, with more to come when we have more freedom to move parameters about. The code in `CreateCostModel` used to have the costing function shapes hard-coded into it and the shapes would have to coincide with those defined in the R code: now R returns the shape to Haskell so that that hard coding is removed, and that was one of the main things I was trying to achieve here. Ultimately the shapes have to be hard coded somewhere (we can't infer them automatically from the benchmarking results) and it seemed best to only do that in one place and I chose to do it in R since that seemed most sensible.

The way I was thinking about having models that depend on the PV was to create a JSON file with the correct cost model types and coefficients for that PV and then commit it and leave it unchanged forever.  If at a later date we decided to change the shape of a costing function then we'd rewrite the R code to include new modelling a modelling code and probably discard the old code:  we wouldn;t need it any more because we would never need to regenerate the JSON files for previous PVs: they're fixed forever and contain all of the information you need recreate the costing function in Haskell. Maybe there's something bad about that plan that hasn't occurred to me though.

**Comment by michaelpj** [comment]
> I'm also keen to keep the use of inline-r to a minimum: I'm not sure how well it's maintained and there may be a danger that it'll stop working altogether at some point. I think we already got stuck with some patched version for a long time.

It's moderately well-maintained, since it's a Tweag project. e.g. it has had commits in the last week. There was a period where it hadn't been updated for a while, though.

**Comment by kwxm** [comment]
Can I merge this?  The important thing is that increases the flexibility of the Haskell models.   It also reworks the R/Haskell interface, but that's only used during cost model generation; it removes some hard-coded numbers that have been annoying me for a while.  We can revisit that part later if need be.

**Comment by michaelpj** [comment]
Sorry, I haven't got around to it, but if you and Roman are happy it's going in the right direction then go for it!

**Comment by effectfully** [comment]
> Can I merge this? 

image
